tests/templates/test_subroutines/test_qsvt.py

```diff
@@ -0,0 +1,329 @@
+
+    def lst_phis(phis):
+        """Used to test queuing in the next test."""
+        return [qml.PCPhase(i, 2, wires=[0, 1]) for i in phis]
+
+    @pytest.mark.parametrize(
+        ("quantum_function", "phi_func", "A", "phis", "results"),
+        [
+            (
+                qfunc,
+                lst_phis,
+                np.array([[0.1, 0.2], [0.3, 0.4]]),
+                np.array([0.2, 0.3]),
+                [
+                    qml.PCPhase(0.2, dim=2, wires=[0]),
+                    qml.PauliX(wires=[0]),
+                    qml.PCPhase(0.3, dim=2, wires=[0]),
+                ],
+            ),
+            (
+                qfunc2,
+                lst_phis,
+                np.array([[0.1, 0.2], [0.3, 0.4]]),
+                np.array([0.1, 0.2]),
+                [
+                    qml.PCPhase(0.1, dim=2, wires=[0]),
+                    qml.prod(qml.PauliX(wires=0), qml.RZ(0.1, wires=0)),
+                    qml.PCPhase(0.2, dim=2, wires=[0]),
+                ],
+            ),
+        ],
+    )
+    def test_queuing_callables(self, quantum_function, phi_func, A, phis, results):
+        """Test that qml.QSVT queues operations correctly when a function is called inside the qnode"""
+
+        with qml.tape.QuantumTape() as tape:
+            qml.QSVT(quantum_function(A), phi_func(phis), wires=[0, 1])
+
+        for idx, val in enumerate(tape.expand().operations):
+            assert val.name == results[idx].name
+            assert val.parameters == results[idx].parameters
+
+    @pytest.mark.parametrize(
+        ("A", "phis", "wires"),
+        [
+            (
+                [[0.1, 0.2], [0.3, 0.4]],
+                [0.1, 0.2, 0.3],
+                [0, 1],
+            )
+        ],
+    )
+    def test_qsvt_grad(self, A, phis, wires):
+        """Test that qml.grad results are the same as finite difference results"""
+
+        @qml.qnode(qml.device("default.qubit", wires=2))
+        def circuit(A, phis):
+            qml.QSVT(
+                qml.BlockEncode(A, wires=[0, 1]),
+                [qml.PCPhase(phi, 2, wires=[0, 1]) for phi in phis],
+                wires=wires,
+            )
+            return qml.expval(qml.PauliZ(wires=0))
+
+        A = np.array([[0.1j, 0.2], [0.3, 0.4]], dtype=complex, requires_grad=True)
+        phis = np.array([0.1, 0.2, 0.3], requires_grad=True)
+        y = circuit(A, phis)
+
+        mat_grad_results, phi_grad_results = qml.grad(circuit)(A, phis)
+
+        manual_mat_results = [
+            (circuit(A + np.array([[0.0001, 0], [0, 0]]), phis) - y) / 0.0001,
+            (circuit(A + np.array([[0, 0.0001], [0, 0]]), phis) - y) / 0.0001,
+            (circuit(A + np.array([[0, 0], [0.0001, 0]]), phis) - y) / 0.0001,
+            (circuit(A + np.array([[0, 0], [0, 0.0001]]), phis) - y) / 0.0001,
+        ]
+
+        for idx, result in enumerate(manual_mat_results):
+            assert np.isclose(result, np.real(mat_grad_results.flatten()[idx]), atol=1e-4)
```

Comment:
`1e-4` is the ultimate precision?

Reply:
^ comes from the finite difference being done above.